File: moxy-compiler/src/main/java/com/omegar/mvp/compiler/pipeline/Pipeline.java
package com.omegar.mvp.compiler.pipeline;

import java.util.ArrayList;
import java.util.List;

import javax.lang.model.element.TypeElement;

/**
 * Created by Anton Knyazev on 03.12.2020.
 */
@SuppressWarnings("rawtypes")
public class Pipeline {
    private static final Context sNoopContext = new Context.Noop();

    private final List<Processor> mProcessors;

    public Pipeline(List<Processor> processors) {
        mProcessors = processors;
    }

    public void start() {
        Context lastContext = sNoopContext;
        for (int i = mProcessors.size() - 1; i >= 0; i--) {
            Processor processor = mProcessors.get(i);
            lastContext = new Context(processor, lastContext);
        }
        lastContext.next(null);
    }

    @SuppressWarnings("unchecked")
    private static class Context implements PipelineContext {

        private final Processor processor;
        private final Context nextContext;

        public Context(Processor processor, Context nextContext) {
            this.processor = processor;
            this.nextContext = nextContext;
        }

        @Override
        public void next(Object nextData) {
            processor.process(nextData, nextContext);
        }

        @Override
        public void finish() {
            processor.finish(nextContext);
        }

        private static class Noop extends Context {

            public Noop() {
                super(null, null);
            }

            @Override
            public void next(Object nextData) {
                // nothing
            }

            @Override
            public void finish() {
                // nothing
            }
        }

    }

    @SuppressWarnings("unchecked")
    public static class Builder<I, O> {

        private final List<Processor> processors = new ArrayList<>();

        public Builder(Publisher<I> publisher) {
            processors.add(publisher);
        }

        public <T> Builder<T, ?> addProcessor(Processor<I, T> processor) {
            processors.add(processor);
            return (Builder<T, ?>) this;
        }

        public Builder<I, O> addValidator(Validator<I> validator) {
            return (Builder<I, O>) addProcessor(validator);
        }

        public Builder<I, O> unique() {
            processors.add(new UniqueValidator<I>());
            return this;
        }

        public Builder<I, O> copyPublishTo(Publisher<I> publisher) {
            return addProcessor(new CopyToPublisherProcessor(publisher));
        }

        public Builder<I, O> copyTypeElementTo(Publisher<TypeElement> publisher) {
            return addProcessor(new CopyTypeElementHolderProcessor(publisher));
        }

        public Pipeline buildPipeline(Receiver<I> receiver) {
            processors.add(receiver);
            return new Pipeline(processors);
        }

    }

}

File: moxy/compiler/src/main/java/com/omegar/mvp/compiler/pipeline/Pipeline.kt
package com.omegar.mvp.compiler.pipeline

import com.omegar.mvp.compiler.entity.TypeElementHolder
import javax.lang.model.element.TypeElement

/**
 * Created by Anton Knyazev on 03.12.2020.
 */
class Pipeline(private val processors: List<Processor<Any?, Any?>>) {

    fun start() {
        val firstContext = processors.foldRight<Processor<Any?, Any?>, Context?>(null) { processor, lastContext ->
            Context(processor, lastContext)
        }
        firstContext?.next(null)
    }

    private class Context(
        private val processor: Processor<Any?, Any?>?,
        private val nextContext: Context?
    ) : PipelineContext<Any?> {

        override fun next(nextData: Any?) {
            processor?.process(nextData, nextContext)
        }

        override fun finish() {
            processor?.finish(nextContext)
        }
    }

    class Builder<I : Any?, O : Any?>(publisher: Publisher<I>) {
        private val processors: MutableList<Processor<out Any?, out Any?>> = ArrayList()

        init {
            processors.add(publisher)
        }

        fun <T : Any?, R : Any?> addProcessor(processor: Processor<I, T>): Builder<T, R> {
            processors.add(processor)
            @Suppress("UNCHECKED_CAST")
            return this as Builder<T, R>
        }

        fun <R : Any?> addValidator(validator: Validator<I>): Builder<I, R> = addProcessor(validator)

        fun uniqueFilter(): Builder<I, O> {
            processors.add(UniqueValidator<I>())
            return this
        }

        fun <R : Any?> copyPublishTo(publisher: Publisher<I>): Builder<I, R> {
            return addProcessor(CopyToPublisherProcessor(publisher))
        }

        fun buildPipeline(receiver: Receiver<I>): Pipeline {
            processors.add(receiver)
            @Suppress("UNCHECKED_CAST")
            return Pipeline(processors as List<Processor<Any?, Any?>>)
        }
    }

}


fun <I : TypeElementHolder, O> Pipeline.Builder<I, O>.copyTypeElementTo(
    publisher: Publisher<TypeElement>
): Pipeline.Builder<I, O> = addProcessor(CopyTypeElementHolderProcessor(publisher))